File: src/main/java/com/vc/medium/L366.scala
package com.vc.medium


class TreeNode366(var _value: Int) {
  var value: Int = _value
  var left: TreeNode366 = null
  var right: TreeNode366 = null
}

object L366 {
  import scala.collection.mutable
  def findLeaves(root: TreeNode366): List[List[Int]] = {
    val map = new mutable.HashMap[Int, mutable.ListBuffer[Int]]()

    def height(root: TreeNode366): Int = {
      if(root == null) return 0
      val h = 1 + Math.max(height(root.left), height(root.right))
      val list = map.getOrElse(h, new mutable.ListBuffer[Int]())
      list += root.value
      map.put(h, list)
      h
    }

    height(root)
    val res = new Array[List[Int]](map.size)
    map.foreach(x => {
      res(x._1 - 1) = x._2.toList
    })
    res.toList
  }
}

File: src/main/java/com/vc/medium/L366.java
package com.vc.medium;

import com.vc.hard.TreeNode;

import java.util.*;

class L366 {
    private List<List<Integer>> res;
    public List<List<Integer>> findLeaves(TreeNode root) {
        this.res = new ArrayList<>();
        height(root);
        return res;
    }

    private int height(TreeNode root) {
        if(root == null) return -1;
        int level = 1 + Math.max(height(root.left), height(root.right));
        if(res.size() < level + 1) res.add(new ArrayList<>());
        res.get(level).add(root.val);
        return level;
    }
}
